File: DesignPattenStudy/DecoratorPattern/Java/Concrete/WaterDecorator.java
package DecoratorPattern.Java.Concrete;

import DecoratorPattern.Java.Abstract.Component;
import DecoratorPattern.Java.Abstract.Decorator;

public class WaterDecorator extends Decorator {

    // americano
    public WaterDecorator(Component component) {

        super(component);
    }

    @Override
    public Integer addPrice() {

        return super.addPrice() + new EspressoDecorator(new BaseComponent()).addPrice() + 500;
    }

    @Override
    public String showItem() {
        return super.showItem() + " + Water : Americano";
    }
}

File: DesignPattenStudy/DecoratorPattern/Kotlin/Concrete/WaterDecorator.kt
package DecoratorPattern.Kotlin.Concrete

import DecoratorPattern.Kotlin.Astract.Component
import DecoratorPattern.Kotlin.Astract.Decorator

class WaterDecorator(override val component: Component) : Decorator(component) {

    override fun addPrice(): Int = super.addPrice() + EspressoDecorator(BaseComponent()).addPrice() + 500

    override fun showItem(): String = "${super.showItem()} + Water : Americano"
}